patroni/quorum.py

```diff
@@ -0,0 +1,336 @@
+            quorum += max(self.numsync - self.numsync_confirmed, 0)
+
+        if (self.leader, quorum, voters) == (old_leader, self.quorum, self.voters):
+            if self.voters:
+                return
+            # If transition produces no change of leader/quorum/voters we want to give a hint to
+            # the caller to fetch the new state from the database and restart QuorumStateResolver.
+            yield 'restart', self.leader, self.quorum, self.voters
+
+        self.quorum = quorum
+        self.voters = voters
+        self.check_invariants()
+        logger.debug('quorum %s %s %s', self.leader, self.quorum, self.voters)
+        yield 'quorum', self.leader, self.quorum, self.voters
+
+    def sync_update(self, numsync: int, sync: CaseInsensitiveSet) -> Iterator[Tuple[str, str, int, CaseInsensitiveSet]]:
+        """Updates numsync and sync fields.
+
+        :param numsync: the new value for `self.numsync`
+        :param sync: the new value for `self.sync`
+        :rtype: Iterator[tuple('sync', leader, numsync, sync)] with the new state of synchronous_standby_names
+        :raises `QuorumError`: in case of invalid data ot if invariant after transition could not be satisfied"""
+
+        if numsync < 0:
+            raise QuorumError("Sync %d < 0 of (%s)" % (numsync, sync))
+        if numsync > 0 and numsync > len(sync):
+            raise QuorumError("Sync %s > N of (%s)" % (numsync, sync))
+
+        self.numsync = numsync
+        self.sync = sync
+        self.check_invariants()
+        logger.debug('sync %s %s %s', self.leader, self.numsync, self.sync)
+        yield 'sync', self.leader, self.numsync, self.sync
+
+    def __iter__(self) -> Iterator[Tuple[str, str, int, CaseInsensitiveSet]]:
+        transitions = list(self._generate_transitions())
+        # Merge 2 transitions of the same type to a single one. This is always safe because skipping the first
+        # transition is equivalent to no one observing the intermediate state.
+        for cur_transition, next_transition in zip(transitions, transitions[1:]+[None]):
+            if next_transition and cur_transition[0] == next_transition[0]:
+                continue
+            yield cur_transition
+            if cur_transition[0] == 'restart':
+                break
+
+    def _generate_transitions(self) -> Iterator[Tuple[str, str, int, CaseInsensitiveSet]]:
+        logger.debug("Quorum state: leader %s quorum %s, voters %s, numsync %s, sync %s, "
+                     "numsync_confirmed %s, active %s, sync_wanted %s leader_wanted %s",
+                     self.leader, self.quorum, self.voters, self.numsync, self.sync,
+                     self.numsync_confirmed, self.active, self.sync_wanted, self.leader_wanted)
+        try:
+            if self.leader_wanted != self.leader:
+                voters = (self.voters - CaseInsensitiveSet([self.leader_wanted])) | CaseInsensitiveSet([self.leader])
```

Comment:
will it make sense to use those `add()` and `discard()` methods here? Actually, the latter seems not to be used anywhere

Reply:
`add()` and `discard()` are mutating existing objects, but we need to create a copy.
These methods are called from the parent class `MutableSet()`.

Reply:
I had `voters` and `self.voters` mixed up in one thing 🤦‍♀️